Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityDeviceManager.cs

```diff
@@ -0,0 +1,164 @@
+                        var controller = GetOrAddController(LastDeviceList[i]);
+
+                        if (controller != null)
+                        {
+                            InputSystem?.RaiseSourceLost(controller.InputSource, controller);
+                        }
+                        ActiveControllers.Remove(LastDeviceList[i]);
+                    }
+                }
+            }
+
+            for (var i = 0; i < joystickNames.Length; i++)
+            {
+                if (string.IsNullOrEmpty(joystickNames[i]))
+                {
+                    continue;
+                }
+
+                if (!ActiveControllers.ContainsKey(joystickNames[i]))
+                {
+                    var controller = GetOrAddController(joystickNames[i]);
+
+                    if (controller != null)
+                    {
+                        InputSystem?.RaiseSourceDetected(controller.InputSource, controller);
+                    }
+                }
+            }
+
+            LastDeviceList = joystickNames;
+        }
+
+        protected virtual GenericUnityController GetOrAddController(string joystickName)
+        {
+            if (ActiveControllers.ContainsKey(joystickName))
+            {
+                var controller = ActiveControllers[joystickName];
+                Debug.Assert(controller != null);
+                return controller;
+            }
+
+            Type controllerType;
+
+            switch (GetCurrentControllerType(joystickName))
+            {
+                default:
+                    return null;
+                case SupportedControllerType.GenericUnityDevice:
+                    controllerType = typeof(GenericUnityController);
+                    break;
+                case SupportedControllerType.XboxController:
+                    controllerType = typeof(XboxUnityController);
+                    break;
+            }
+
+            var inputSource = InputSystem?.RequestNewGenericInputSource($"{controllerType.Name} Controller");
+            var detectedController = Activator.CreateInstance(controllerType, TrackingState.NotTracked, Handedness.None, inputSource, null) as GenericUnityController;
+            if (detectedController == null) { Debug.LogError($"Failed to create {controllerType.Name} controller"); }
+            detectedController?.SetupConfiguration(controllerType);
+
+            Debug.Log($"Found controller: {joystickName}");
+            ActiveControllers.Add(joystickName, detectedController);
+            return detectedController;
+        }
+
+        protected virtual SupportedControllerType GetCurrentControllerType(string joystickName)
+        {
+            if (string.IsNullOrEmpty(joystickName) ||
+                joystickName.Contains("OpenVR") ||
+                joystickName.Contains("Spatial"))
+            {
+                return SupportedControllerType.None;
+            }
+
+            if (joystickName.Contains("Xbox Controller") ||
```

Comment:
Why not just use "Xbox"?  Seems a little convoluted.

Reply:
Ad what about generic gamepads?  They wouldn't be supported here.

Reply:
Because Unity names the joysticks by specific strings, anything else would fail to recognize the xbox controller by this joystick name.

Reply:
> And what about generic gamepads? They wouldn't be supported here.

That's not true, anything else falls back to being a generic unity controller.